cmd/uplink/cmd/share.go

```diff
@@ -0,0 +1,166 @@
+		RunE:  shareMain,
+	}, RootCmd)
+
+	cfgstruct.Bind(shareCmd.Flags(), &shareCfg)
+}
+
+const shareISO8601 = "2006-01-02T15:04:05-0700"
+
+func parseHumanDate(date string, now time.Time) (*time.Time, error) {
+	if date == "" {
+		return nil, nil
+	} else if date == "now" {
+		return &now, nil
+	} else if date[0] == '+' {
+		d, err := time.ParseDuration(date[1:])
+		t := now.Add(d)
+		return &t, errs.Wrap(err)
+	} else if date[0] == '-' {
+		d, err := time.ParseDuration(date[1:])
+		t := now.Add(-d)
+		return &t, errs.Wrap(err)
+	} else {
+		t, err := time.Parse(shareISO8601, date)
+		return &t, errs.Wrap(err)
+	}
+}
+
+// shareMain is the function executed when shareCmd is called
+func shareMain(cmd *cobra.Command, args []string) (err error) {
+	ctx := process.Ctx(cmd)
+	now := time.Now()
+
+	notBefore, err := parseHumanDate(shareCfg.NotBefore, now)
+	if err != nil {
+		return err
+	}
+	notAfter, err := parseHumanDate(shareCfg.NotAfter, now)
+	if err != nil {
+		return err
+	}
+
+	// TODO(jeff): we have to have the server side of things expecting macaroons
+	// before we can change libuplink to use macaroons because of all the tests.
+	// For now, just use the raw macaroon library.
+
+	key, err := macaroon.ParseAPIKey(cfg.Client.APIKey)
+	if err != nil {
+		return err
+	}
+
+	caveat := macaroon.NewCaveat()
+	caveat.DisallowDeletes = shareCfg.DisallowDeletes || shareCfg.Readonly
+	caveat.DisallowLists = shareCfg.DisallowLists || shareCfg.Writeonly
+	caveat.DisallowReads = shareCfg.DisallowReads || shareCfg.Writeonly
+	caveat.DisallowWrites = shareCfg.DisallowWrites || shareCfg.Readonly
+	caveat.NotBefore = notBefore
+	caveat.NotAfter = notAfter
+
+	var project *libuplink.Project
+	var access libuplink.EncryptionAccess
+	copy(access.Key[:], []byte(cfg.Enc.Key))
+	cache := make(map[string]*libuplink.BucketConfig)
+
+	for _, path := range shareCfg.AllowedPathPrefix {
+		p, err := fpath.New(path)
+		if err != nil {
+			return err
+		}
+		if p.IsLocal() {
+			return errs.New("required path must be remote: %q", path)
+		}
+
+		bi, ok := cache[p.Bucket()]
+		if !ok {
+			if project == nil {
+				project, err = cfg.GetProject(ctx)
+				if err != nil {
+					return err
+				}
+				defer func() { err = errs.Combine(err, project.Close()) }()
+			}
+
+			_, bi, err = project.GetBucketInfo(ctx, p.Bucket())
+			if err != nil {
+				return err
+			}
+
+			cache[p.Bucket()] = bi
+		}
+
+		encPath, err := encryption.EncryptPath(path, bi.PathCipher.ToCipher(), &access.Key)
+		if err != nil {
+			return err
+		}
+
+		caveat.AllowedPaths = append(caveat.AllowedPaths, &macaroon.Caveat_Path{
+			Bucket:              []byte(p.Bucket()),
+			EncryptedPathPrefix: []byte(encPath),
+		})
+	}
+
+	{
+		// Times don't marshal very well with MarshalTextString, and the nonce doesn't
+		// matter to humans, so handle those explicitly and then dispatch to the generic
+		// routine to avoid having to print all the things individually.
+		caveat := caveat
```

Comment:
i wish there was a way to tell go that you really do mean to make a copy here and if caveat ever becomes a pointer that it should yell

Reply:
maybe this stanza could be a helper function to make it more clear what's going on

Reply:
I'll change it to `proto.Clone` the thing.